File: Python/cam-v4l2-dual/main.py
#
# Dual video capture in OpenCV with V4L2
#

import cv2 as cv
import numpy as np
import sys

# get the first video device node
camera_arg_1 = sys.argv[1][10:]

# get the second video device node
camera_arg_2 = sys.argv[2][10:]

# Select v4l2 for video capture
capture_1 = cv.VideoCapture(int(camera_arg_1), cv.CAP_V4L2)
capture_2 = cv.VideoCapture(int(camera_arg_2), cv.CAP_V4L2)

# Configure video format (NV12) and Resolution (3840x2160)
capture_1.set(cv.CAP_PROP_FOURCC, cv.VideoWriter_fourcc('U', 'Y', 'V', 'Y'))
capture_1.set(cv.CAP_PROP_FRAME_WIDTH, 3840)
capture_1.set(cv.CAP_PROP_FRAME_HEIGHT, 2160)

capture_2.set(cv.CAP_PROP_FOURCC, cv.VideoWriter_fourcc('U', 'Y', 'V', 'Y'))
capture_2.set(cv.CAP_PROP_FRAME_WIDTH, 3840)
capture_2.set(cv.CAP_PROP_FRAME_HEIGHT, 2160)


cv.namedWindow("video 1", cv.WINDOW_NORMAL)
cv.namedWindow("video 2", cv.WINDOW_NORMAL)

# Preview 100 Video frames at ~30 Frames per second
for _ in range(100):
    ret, frame_1 = capture_1.read()
    ret, frame_2 = capture_2.read()

    if not ret or frame_1 is None or frame_2 is None:
        break

    cv.imshow("video 1", frame_1)
    cv.imshow("video 2", frame_2)
    cv.waitKey(30)

capture_1.release()
capture_2.release()
cv.destroyAllWindows()

File: C++/cam-v4l2-dual/main.cpp
#include <iostream>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/core/core.hpp>

int main(int argc, char* argv[]) {
    // Get the video device node
    std::string camera_arg_1 = argv[1];
    std::string camera_arg_2 = argv[2];

    // Capture Parameters
    int width = 3840;
    int height = 2160;
    int fourcc = cv::VideoWriter::fourcc('U', 'Y', 'V', 'Y');

    // Select v4l2 for video capture
    cv::VideoCapture capture_1(std::stoi(camera_arg_1.substr(10)), cv::CAP_V4L2);
    cv::VideoCapture capture_2(std::stoi(camera_arg_2.substr(10)), cv::CAP_V4L2);

    // Check if the video capture is successful
    if (!capture_1.isOpened()) {
        std::cout << "Error opening video 1 capture." << std::endl;
        return 1;
    }

    if (!capture_2.isOpened()) {
        std::cout << "Error opening video 2 capture." << std::endl;
        return 1;
    }

    // Configure video format and resolution
    capture_1.set(cv::CAP_PROP_FOURCC, fourcc);
    capture_1.set(cv::CAP_PROP_FRAME_WIDTH, width);
    capture_1.set(cv::CAP_PROP_FRAME_HEIGHT, height);

    capture_2.set(cv::CAP_PROP_FOURCC, fourcc);
    capture_2.set(cv::CAP_PROP_FRAME_WIDTH, width);
    capture_2.set(cv::CAP_PROP_FRAME_HEIGHT, height);

    // Create a window to view the video
    cv::namedWindow("video 1", cv::WINDOW_NORMAL);
    cv::namedWindow("video 2", cv::WINDOW_NORMAL);

    // Preview 100 video frames at ~30 frames per second
    for (int i = 0; i < 100; ++i) {
        cv::Mat frame_1, frame_2;
        capture_1 >> frame_1;
        capture_2 >> frame_2;

        if (frame_1.empty() || frame_2.empty()) {
            break;
        }

        cv::imshow("video 1", frame);
        cv::imshow("video 2", frame);
        cv::waitKey(30);
    }

    // Release the video capture
    capture_1.release();
    capture_2.release();
    cv::destroyAllWindows();

    return 0;
}
